src/fd/eventfd.rs

```diff
@@ -0,0 +1,202 @@
+		Self {
+			state: Mutex::new(EventState::new(counter)),
+			flags: self.flags,
+		}
+	}
+}
+
+impl EventFd {
+	pub fn new(initval: u64, flags: EventFlags) -> Self {
+		debug!("Create EventFd {}, {:?}", initval, flags);
+		Self {
+			state: Mutex::new(EventState::new(initval)),
+			flags,
+		}
+	}
+}
+
+#[async_trait]
+impl ObjectInterface for EventFd {
+	async fn async_read(&self, buf: &mut [u8]) -> Result<usize, IoError> {
+		let len = mem::size_of::<u64>();
+
+		if buf.len() < len {
+			return Err(IoError::EINVAL);
+		}
+
+		future::poll_fn(|cx| {
+			if self.flags.contains(EventFlags::EFD_SEMAPHORE) {
+				let mut pinned = core::pin::pin!(self.state.lock());
+				if let Poll::Ready(mut guard) = pinned.as_mut().poll(cx) {
+					if guard.counter > 0 {
+						guard.counter -= 1;
+						let tmp = u64::to_ne_bytes(1);
+						buf[..len].copy_from_slice(&tmp);
+						if let Some(cx) = guard.write_queue.pop_front() {
+							cx.wake_by_ref();
+						}
+						Poll::Ready(Ok(len))
+					} else {
+						guard.read_queue.push_back(cx.waker().clone());
+						Poll::Pending
+					}
+				} else {
+					Poll::Pending
+				}
+			} else {
+				let mut pinned = core::pin::pin!(self.state.lock());
+				if let Poll::Ready(mut guard) = pinned.as_mut().poll(cx) {
+					let tmp = guard.counter;
+					if tmp > 0 {
+						guard.counter = 0;
+						buf[..len].copy_from_slice(&u64::to_ne_bytes(tmp));
+						if let Some(cx) = guard.read_queue.pop_front() {
+							cx.wake_by_ref();
+						}
+						Poll::Ready(Ok(len))
+					} else {
+						guard.read_queue.push_back(cx.waker().clone());
+						Poll::Pending
+					}
+				} else {
+					Poll::Pending
+				}
+			}
+		})
+		.await
+	}
+
+	async fn async_write(&self, buf: &[u8]) -> Result<usize, IoError> {
+		let len = mem::size_of::<u64>();
+
+		if buf.len() < len {
+			return Err(IoError::EINVAL);
+		}
+
+		let c = u64::from_ne_bytes(buf[..len].try_into().unwrap());
+
+		future::poll_fn(|cx| {
+			let mut pinned = core::pin::pin!(self.state.lock());
+			if let Poll::Ready(mut guard) = pinned.as_mut().poll(cx) {
+				if u64::MAX - guard.counter > c {
+					guard.counter += c;
+					if self.flags.contains(EventFlags::EFD_SEMAPHORE) {
+						for _i in 0..c {
+							if let Some(cx) = guard.read_queue.pop_front() {
+								cx.wake_by_ref();
+							} else {
+								break;
+							}
+						}
+					} else if let Some(cx) = guard.read_queue.pop_front() {
+						cx.wake_by_ref();
+					}
+
+					Poll::Ready(Ok(len))
+				} else {
+					guard.write_queue.push_back(cx.waker().clone());
+					Poll::Pending
+				}
+			} else {
+				Poll::Pending
+			}
+		})
+		.await
+	}
+
+	async fn poll(&self, event: PollEvent) -> Result<PollEvent, IoError> {
+		let mut result: PollEvent = PollEvent::empty();
+		let guard = self.state.lock().await;
+
+		if guard.counter < u64::MAX - 1 {
+			if event.contains(PollEvent::POLLOUT) {
+				result.insert(PollEvent::POLLOUT);
+			}
```

Comment:
This could be heavily deduplicated by using a closure. Again, I can do this after merging this PR or before.